python/tank_vendor/shotgun_deploy/descriptor.py

```diff
@@ -0,0 +1,286 @@
+        return EngineDescriptor(io_descriptor)
+
+    elif descriptor_type == Descriptor.FRAMEWORK:
+        return FrameworkDescriptor(io_descriptor)
+
+    elif descriptor_type == Descriptor.CONFIG:
+        return ConfigDescriptor(io_descriptor)
+
+    elif descriptor_type == Descriptor.CORE:
+        return CoreDescriptor(io_descriptor)
+
+    else:
+        raise ShotgunDeployError("Unsupported descriptor type %s" % descriptor_type)
+
+
+class Descriptor(object):
+    """
+    A descriptor describes a particular version of an app, engine or core component.
+    It also knows how to access metadata such as documentation, descriptions etc.
+
+    Descriptor is subclassed to distinguish different types of payload;
+    apps, engines, configs, cores etc. Each payload may have different accessors
+    and helper methods.
+    """
+
+    (APP, FRAMEWORK, ENGINE, CONFIG, CORE) = range(5)
+
+    def __init__(self, io_descriptor):
+        """
+        Constructor
+
+        :param io_descriptor: Associated IO descriptor.
+        """
+        # construct a suitable IO descriptor for this locator
+        self._io_descriptor = io_descriptor
+
+    def __eq__(self, other):
+        # By default, we can assume equality if the path to the data
+        # on disk is equivalent.
+        if isinstance(other, self.__class__):
+            return self.get_path() == other.get_path()
+        else:
+            return False
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __repr__(self):
+        """
+        Used for debug logging
+        """
+        class_name = self.__class__.__name__
+        return "<%s %r>" % (class_name, self._io_descriptor)
+
+    def __str__(self):
+        """
+        Used for pretty printing
+        """
+        return "%s %s" % (self.get_system_name(), self.get_version())
+
+    ###############################################################################################
+    # data accessors
+
+    def get_location(self):
+        """
+        Returns the location dict associated with this descriptor
+
+        :returns: Dictionary that can be used to construct the descriptor
+        """
+        return self._io_descriptor.get_location()
+
+    def copy(self, target_folder):
+        """
+        Copy the config descriptor into the specified target location
+
+        :param target_folder: Folder to copy the descriptor to
+        """
+        self._io_descriptor.copy(target_folder)
+
+    def get_display_name(self):
+        """
+        Returns the display name for this item.
+        If no display name has been defined, the system name will be returned.
+
+        :returns: Display name as string
+        """
+        meta = self._io_descriptor.get_manifest()
+        display_name = meta.get("display_name")
+        if display_name is None:
+            display_name = self.get_system_name()
+        return display_name
+
+    def is_developer(self):
```

Comment:
Great points. @jfboismenu - there are several modes - developer doesn't mean local, it really means developer - this is to make it clearer when to enable stuff like reload menus etc. So the dev descriptor is a "developer" mode where the "path" descriptor isn't. Similarly, there is a "immutable" flag to indicate that a descriptor will never be manually edited. @josh-t i'll change to `is_dev` - i was thinking "developer *mode*" rather than "i am the developer!". Dev is better, reflects the name of the descriptor too.